### PR TITLE
Node_Animation Model Group

### LINKED ISSUE #404: Make sure to cover this case when implementing animation tests
https://github.com/BabylonJS/Babylon.js/issues/4224

### PR DESCRIPTION
**Pull request by stevk**
Implements the Node_Animation model group.

Changed some of the runtime animation objects from structs to classes because the structs were causing issues with how we've been passing references.

Addresses several parts of #414 
Fixes #404, #428 

TODO:
- [x] I'm not sure I've correctly implemented a model for animating a curve that doesn't start at 0.
- [x] A proper texture still needs to be created to show which part of the model is supposed to be facing which way. 
- [ ] The validator shows an error in the bufferView for "Animation Sampler Output" for all of the models that are doing rotations. I'm not sure of the exact cause. "Accessor element at index # is not of unit length: 2."
- [ ] I'm sure we can pick better values for the animations that what is currently plugged in. 
- [ ] Maybe we should add a notes column to describe the expected animation for each model?

### CODE REVIEW

File: Output/Animation_Node/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+These models are intended to test the basic attributes of animation on a node.  
+
+The last model on this list creates a curve which doesn't start at Zero.  
+
+The following table shows the properties that are set for a given model.  
+
+|   | Sample Image | Translation | Rotation | Scale | Interpolation |
```
Comment (bghgary) [inline Output/Animation_Node/README.md]: Let's try a single column for `Target(s)` and list the targets.

File: Source/Runtime/Animation.cs [hunk 2]
```diff
@@ -4,9 +4,9 @@
 
 namespace AssetGenerator.Runtime
 {
-    internal struct Animation
+    internal class Animation
     {
         public String Name { get; set; }
-        public List<AnimationChannel> AnimationChannels { get; set; }
```
Comment (bghgary) [inline Source/Runtime/Animation.cs]: Restore the `Animation` prefix for types.

File: Source/ModelGroups/Animation_Node.cs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+                properties.Add(new Property(PropertyName.Interpolation, "Cubic Spline"));
+            }
+
+            void CreateMultipleChannelsWithUniqueTargets(List<Property> properties, List<Runtime.Channel> channels, Runtime.Node node)
+            {
+                // The first channel is already added as a common property.
+                channels.Add(new Runtime.Channel());
+                channels.Add(new Runtime.Channel());
+
+                SetTranslationChannelTarget(properties, channels[0], node);
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: create a new properties list and then combine the strings afterwards into the real properties list

File: Source/Runtime/Extensions/KHR_texture_transform.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+﻿using System.Numerics;
```
Comment (bghgary) [inline Source/Runtime/Extensions/KHR_texture_transform.cs]: oops, i accidentally put this in. please remove this.

File: Source/ModelGroup_Cube.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        protected static partial class MeshPrimitive
+        {
+            public static Runtime.MeshPrimitive CreateCube()
+            {
+                return new Runtime.MeshPrimitive
+                {
+                    Positions = new List<Vector3>()
+                    {
+                        // Left
+                        new Vector3( -0.5f, -0.5f, -0.5f),
```
Comment (bghgary) [inline Source/ModelGroup_Cube.cs]: Scale to 0.1

File: Source/ModelGroups/Animation_Node.cs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+                };
+                properties.Add(new Property(PropertyName.Scale, ":white_check_mark:"));
+            }
+
+            void SetLinearSampler(List<Property> properties, Runtime.Channel channel)
+            {
+                channel.Sampler = new Runtime.LinearAnimationSampler<Vector3>(
+                    new List<float>
+                    {
+                        0.0f,
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: 0.0, 1.0, 2.0 to all

File: Source/ModelGroups/Animation_Node.cs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+                        0.5f,
+                        1.0f,
+                    },
+                    new List<Vector3>
+                    {
+                        new Vector3(0, 0, 0),
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: (-0.5, 0.0, 0.0), (0.0, 0.0, 0.0), (0.5, 0.0, 0.0)

File: Source/ModelGroup_Cube.cs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                            new Vector2(0.500f, 0.000f),
+                            new Vector2(0.000f, 0.000f),
+
+                            // Front
+                            new Vector2(0.500f, 0.333f),
+                            new Vector2(1.000f, 0.333f),
+                            new Vector2(1.000f, 0.000f),
+                            new Vector2(0.500f, 0.000f),
+
+                            // Right
```
Comment (bghgary) [inline Source/ModelGroup_Cube.cs]: Left <-> Right

File: Source/ModelGroups/Animation_Node.cs [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+                channel.Sampler = new Runtime.LinearAnimationSampler<Quaternion>(
+                    new List<float>
+                    {
+                        0.0f,
+                        0.5f,
+                        1.0f,
+                    },
+                    new List<Quaternion>
+                    {
+                        new Quaternion(0, 0, 0, 0),
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: Show Right, Front, then Left

File: Source/ModelGroups/Animation_Node.cs [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+            {
+                channel.Target = new Runtime.ChannelTarget
+                {
+                    Node = node,
+                    Path = Runtime.ChannelTarget.PathEnum.SCALE,
+                };
+                properties.Add(new Property(PropertyName.Scale, ":white_check_mark:"));
+            }
+
+            void SetLinearSampler(List<Property> properties, Runtime.Channel channel)
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: Create a new function for Scale and rename this to be specific for Translation. Make scale half, normal, double

File: Source/ModelGroups/Animation_Node.cs [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+                        {
+                            InTangent = new Vector3(1, 1, 1),
+                            Value = new Vector3(0.1f, 0.1f, 0.1f),
+                            OutTangent = new Vector3(1,1, 1)
+                        },
+                        new Runtime.CubicSplineSampler<Vector3>.Key
+                        {
+                            InTangent = new Vector3(1, 1, 1),
+                            Value = new Vector3(0.1f, 0.1f, 0.1f),
+                            OutTangent = new Vector3(1, 1, 1)
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: Change to zero.

File: Source/ModelGroups/Animation_Node.cs [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+                        0.0f,
+                        0.5f,
+                        1.0f,
+                    },
+                    new List<Runtime.CubicSplineSampler<Vector3>.Key>
+                    {
+                        new Runtime.CubicSplineSampler<Vector3>.Key
+                        {
+                            InTangent = new Vector3(0, 0, 0),
+                            Value = new Vector3(1, 1, 1),
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: -0.5, 0.0, 0.0

File: Output/Animation_Node/README.md (hunk 1 above)
Comment (stevk) [reply Output/Animation_Node/README.md]: The `(s)` causes an error when used as a enum, so I went with `Targets` instead.

Comment (stevk) [reply Output/Animation_Node/README.md]: Fixed

File: Source/ModelGroups/Animation_Node.cs (hunk 3 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: Fixed

File: Source/Runtime/Animation.cs (hunk 2 above)
Comment (stevk) [reply Source/Runtime/Animation.cs]: Fixed

File: Source/Runtime/Extensions/KHR_texture_transform.cs (hunk 4 above)
Comment (stevk) [reply Source/Runtime/Extensions/KHR_texture_transform.cs]: Fixed

File: Source/ModelGroups/Animation_Node.cs (hunk 10 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: Fixed

File: Source/ModelGroups/Animation_Node.cs (hunk 6 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: Fixed

File: Source/ModelGroups/Animation_Node.cs (hunk 11 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: Fixed

File: Source/ModelGroup_Cube.cs (hunk 8 above)
Comment (stevk) [reply Source/ModelGroup_Cube.cs]: Fixed

File: Source/ModelGroups/Animation_Node.cs (hunk 7 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: Fixed

File: Source/ModelGroup_Cube.cs (hunk 5 above)
Comment (stevk) [reply Source/ModelGroup_Cube.cs]: Fixed

File: Source/ModelGroups/Animation_Node.cs (hunk 6 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: This left a jump at the end of the animation. I believe it looks better with a full loop, so I opted for the following for a repeating Left > Right > Left motion:

```
channel.Sampler = new Runtime.LinearAnimationSampler<Vector3>(
                    new List<float>
                    {
                        0.0f,
                        2.0f,
                        4.0f,
                    },
                    new List<Vector3>
                    {
                        new Vector3(-0.5f, 0.0f, 0.0f),
                        new Vector3(0.5f, 0.0f, 0.0f),
                        new Vector3(-0.5f, 0.0f, 0.0f),
                    });
```

File: Source/ModelGroups/Animation_Node.cs (hunk 10 above)
Comment (stevk) [reply Source/ModelGroups/Animation_Node.cs]: I made scale loop as well. Is it really nessessary to specify the 0,0,0 value? As far as I can tell, setting that middle value doesn't make a difference due to being exactly between the min/max values with equal time spacing between.

**Comment by stevk** [comment]
Looking into #428 tomorrow. The model also appears to change scale when rotating, but I'm not sure if that is an issue or not.

File: Source/ModelGroups/Animation_Node.cs [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+                            InTangent = new Vector3(1, 1, 1),
+                            Value = new Vector3(0.1f, 0.1f, 0.1f),
+                            OutTangent = new Vector3(1, 1, 1)
+                        }
+                    });
+
+                properties.Add(new Property(PropertyName.Interpolation, "Cubic Spline"));
+            }
+
+            void SetLinearSamplerForRotation(List<Property> properties, Runtime.AnimationChannel channel)
```
Comment (bghgary) [inline Source/ModelGroups/Animation_Node.cs]: Move this with the linear functions

File: Output/Animation_Node/README.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
+These models are intended to test the basic attributes of animation on a node.  
+
+The last model on this list creates a curve which doesn't start at Zero.
```
Comment (bghgary) [inline Output/Animation_Node/README.md]: Remove this

File: Source/Runtime/AnimationChannelTarget.cs [hunk 15]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace AssetGenerator.Runtime
 {
-    internal struct AnimationChannelTarget
+    internal class ChannelTarget
```
Comment (bghgary) [inline Source/Runtime/AnimationChannelTarget.cs]: Revert this

File: Source/Property.cs [hunk 16]
```diff
@@ -142,6 +142,8 @@ internal enum PropertyName
         Translation,
         Rotation,
         Scale,
+        Targets,
```
Comment (bghgary) [inline Source/Property.cs]: Change to singular

Comment (stevk) [reply Source/Property.cs]: Fixed

File: Source/Runtime/AnimationChannelTarget.cs (hunk 15 above)
Comment (stevk) [reply Source/Runtime/AnimationChannelTarget.cs]: Fixed

File: Output/Animation_Node/README.md (hunk 14 above)
Comment (stevk) [reply Output/Animation_Node/README.md]: Fixed

File: Source/Runtime/AnimationChannelTarget.cs (hunk 15 above)
Comment (bghgary) [reply Source/Runtime/AnimationChannelTarget.cs]: I meant the name and not the class to struct change.

Comment (stevk) [reply Source/Runtime/AnimationChannelTarget.cs]: Oops, switched that back around.

**Pull request merged**